test(doctest): launch doctests in Travis for README & main package

ankostis opened the pull request:
Use `nosetests` command to launch `--with-doctest` and check main sources and `README.rst` file.

- Still not all docs pass the tests.
+ Install `corner` package on travis used by test-sample.

ankostis commented:
For some reason, model.git() fails in 4 calls from `/doc/model.rst` doctests.
+ Also `lmfit.model` and its subpackages is still excluded from doctesting (many failures).

newville commented:
@ankostis I'm sort of -0.01 on relying on doctests: it's fine to fix places where the examples in the docstrings are incorrect, but encouraging the use of docstring examples as tests seems less important to me.  That is, we want to see actual tests, with assertions, expected results, expected failures, etc.   

Docstrings are first and foremost documentation for humans to read.  If they can be automatically verified for correctness without hindering readability, fine.  If that cannot be done (for example if extra boilerplate or obvious imports are needed), it hinders the documentation in favor of the test.
If something like `# doctest: SKIP` is needed, that is not improving the documentation.

More importantly, I think we want to not depend heavily on nose, and should be migrating to pytest instead.

ankostis commented:
That's a lot of effort on my part (many hours) to waste it.
But it's your call.

andyfaff commented:
It's great that you're contributing to making lmfit better for everyone.
However, as with many open source projects it's often a good idea to
propose changes to the mailing list before a lot of time is spent on
changes that may not get integrated.
For me doctests make sure that a typical user should be able to run them
themselves. However, they need to be really clear and I agree with Matt
that cruft like # doctest: SKIP shouldn't be seen.

On 21 June 2017 at 17:04, Kostis Anagnostopoulos <notifications@github.com>
wrote:

> That's a lot of effort on my part (many hours) to waste it.
> But it's your call.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/lmfit/lmfit-py/pull/419#issuecomment-309982889>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAq51i20EuO09keyv_Mexosc-8oHQNKkks5sGMCLgaJpZM4OANgZ>
> .
>



-- 
_____________________________________
Dr. Andrew Nelson


_____________________________________

ankostis commented:
> I agree with Matt that cruft like # doctest: SKIP shouldn't be seen.# doctest: SKIP shouldn't be seen.

You can count me in on that.  
That's why I left the 2nd commit to fail, so you can see why it is failing and fix it - I couldn't.
In addition:

- I've added only 3(!) `+SKIP` directives, on the `pyplot.plot()` invocations, to avoid halting doctests when a matplotlib diagram pops-up.  Is that ok?
- There are some additional skips in the `nosetest` command for the `lmfit.model` package because it failed in too-many places that I could not resolve - maybe the doctests in the code were from an older era, no?  
  Anyhow, I've already asked for help on that, no my 2nd comment.
- For all the rest, I've fixed the errors, added deterministic random keys, retrofitted comparisons, limited changes, etc, to make them pass decently.

I've never ever met an open-source project where augmenting *tests* is not met with applaud.
I believe here also there won't be an exception :-)

ankostis commented:
Regarding the phasing-out of nosetests: good.  
There are very few changes related to *nose* in this PR, and are all isolated in the `travis.yaml` file.  
So in my view , doctesting and moving from *nose-->pytest* remain decoupled.

ankostis commented:
I don't feel like contributing to Mat; he is too proud to admit he has it wrong...

(luckily his colleagues [wake him, up sometimes](/lmfit/lmfit-py/issues/149#issuecomment-62584117))
